coreneuron/io/phase2.cpp

```diff
@@ -0,0 +1,1211 @@
+            ml->pdata = (int*)ecalloc_align(nrn_soa_padded_size(n, layout) * szdp, sizeof(int));
+            std::copy(tmls[itml].pdata.begin(), tmls[itml].pdata.end(), ml->pdata);
+            mech_layout<int>(ml->pdata, n, szdp, layout);
+
+#if CHKPNTDEBUG  // Not substantive. Only for debugging.
+            Memb_list_ckpnt* mlc = ntc.mlmap[type];
+            mlc->pdata_not_permuted = (int*)coreneuron::ecalloc_align(n * szdp, sizeof(int));
+            if (layout == 1) {  // AoS just copy
+                for (int i = 0; i < n; ++i) {
+                    for (int j = 0; j < szdp; ++j) {
+                        mlc->pdata_not_permuted[i * szdp + j] = ml->pdata[i * szdp + j];
+                    }
+                }
+            } else if (layout == 0) {  // SoA transpose and unpad
+                int align_cnt = nrn_soa_padded_size(n, layout);
+                for (int i = 0; i < n; ++i) {
+                    for (int j = 0; j < szdp; ++j) {
+                        mlc->pdata_not_permuted[i * szdp + j] = ml->pdata[i + j * align_cnt];
+                    }
+                }
+            }
+#endif
+        } else {
+            ml->pdata = nullptr;
+        }
+        if (corenrn.get_pnt_map()[type] > 0) {  // POINT_PROCESS mechanism including acell
+            int cnt = ml->nodecount;
+            Point_process* pnt = nullptr;
+            pnt = nt.pntprocs + synoffset;
+            pnt_offset[type] = synoffset;
+            synoffset += cnt;
+            for (int i = 0; i < cnt; ++i) {
+                Point_process* pp = pnt + i;
+                pp->_type = type;
+                pp->_i_instance = i;
+                nt._vdata[ml->pdata[nrn_i_layout(i, cnt, 1, szdp, layout)]] = pp;
+                pp->_tid = nt.id;
+            }
+        }
+    }
+
+    if (nrn_have_gaps) {
+        nrn_partrans::gap_thread_setup(nt);
+    }
+
+    // Some pdata may index into data which has been reordered from AoS to
+    // SoA. The four possibilities are if semantics is -1 (area), -5 (pointer),
+    // -9 (diam), // or 0-999 (ion variables).
+    // Note that pdata has a layout and the // type block in nt.data into which
+    // it indexes, has a layout.
+    for (auto tml = nt.tml; tml; tml = tml->next) {
+        int type = tml->index;
+        int layout = corenrn.get_mech_data_layout()[type];
+        int* pdata = tml->ml->pdata;
+        int cnt = tml->ml->nodecount;
+        int szdp = nrn_prop_dparam_size_[type];
+        int* semantics = memb_func[type].dparam_semantics;
+
+        // ignore ARTIFICIAL_CELL (has useless area pointer with semantics=-1)
+        if (corenrn.get_is_artificial()[type]) {
+            continue;
+        }
+
+        if (szdp) {
+            if (!semantics)
+                continue;  // temporary for HDFReport, Binreport which will be skipped in
+                           // bbcore_write of HBPNeuron
+            nrn_assert(semantics);
+        }
+
+        for (int i = 0; i < szdp; ++i) {
```

Comment:
can we not iterate over semantics? And also use a switch case

Reply:
We need the iterator to pass to some function bellow (`pdata_relocation` and `nrn_i_layout`)

Reply:
More over, semantics is a `int*` not null terminated, do you have any idea how to iterate on it?